### ISSUE TITLE
Saving files on MacOS X

### ISSUE THREAD

**Issue by berlincount** [TB-throwback/LookOut-fix-version#29]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest version 1.2.23
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed (correctly .. #7)

## Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* Operating System: MacOS X 10.14.2
* Thunderbird version number: 60.4.0
* 64 bit/32 bit: 64bit
* LookOut (fix version) version number: 1.2.23

# Expected Behavior

When requesting to save a file, file is saved to the selected location.

# Current Behavior

File is not saved, or saved somewhere else.

## Steps to Reproduce / Failure Information (for bugs)

Please provide detailed steps for reproducing the issue.

Got a windows mail forwarded, with attached PDF. PDF is visible in Attachment list, and can be opened in Preview.app via double-click. Saving after double-click or with right click does not work.

1. right-click file in Attachment-List, select "Save As ..."
2. select place to save in dialog, click Save
3. don't find file saved.

```
LookOut: Entering save_attachment()
LookOut:    got attachment = ({contentType:"application/binary", name:"red\xE4cted.PDF", uri:"imap-message://redacted@imap.redacted/redacted#1234", isExternalAttachment:true, size:97475, partID:"1.2", url:"imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2", lo_orig_open:(function AttachmentInfo_open()
  {
    if (!this.hasFile)
      return;

    if (this.isEmpty) {
      var prompt = document.getElementById("bundle_messenger")
                           .getString("emptyAttachment");
      msgWindow.promptDialog.alert(null, prompt);
    } else {
      messenger.openAttachment(this.contentType, this.url,
                               encodeURIComponent(this.name),
                               this.uri, this.isExternalAttachment);
    }
  }), open:(function () {

				lookout_lib.open_attachment( this );

			}), lo_orig_save:(function AttachmentInfo_save()
  {
    messenger.saveAttachment(this.contentType, this.url,
                             encodeURIComponent(this.name),
                             this.uri, this.isExternalAttachment);
  }), save:(function (save_dir) {

				lookout_lib.save_attachment( this, save_dir );

			}), parent:{contentType:"application/ms-tnef", name:"winmail.dat", uri:"imap-message://redacted@imap.redacted/redacted#1234", isExternalAttachment:false, size:106536, partID:"1.2", url:"imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat"}, part_id:2})
LookOut:   Parent: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat
         Content-Type: application/binary
         Displayname: redäcted.PDF
         Part_ID: 2
         isExternal: true
         URL: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2
         mMsgUri: imap-message://redacted@imap.redacted/redacted#1234
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
TNEF Key: 0803
checksum = 1  sum = 1
checksum = 232  sum = 232
TNEF: OEM Code Page = 1252
checksum = 2097  sum = 2097
checksum = 43593  sum = 43593
TNEF: Entering tnef_file_notify()
TNEF: Notifying listener of body_part_0.rtf, pos = 0, has data = true, is final = true
LookOut: Entering onTnefStart()
LookOut:    open or save: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.1
LookOut:    Parent: null
            mMsgUri: imap-message://redacted@imap.redacted/redacted#1234
            requested Part_ID: 2
            Part_ID: 1
            Displayname: body_part_0.rtf
            Content-Type: application/rtf
            Length: 14646
            URL: 
            mimeurl: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.1
LookOut: onTnefData position 0  data.len 14646  outstrm null
LookOut: Entering onTnefEnd()
Lookout: Entering redraw_attachment_view()
checksum = 1085  sum = 1085
starting file sub-attachment
checksum = 317  sum = 317
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
checksum = 39746  sum = 39746
checksum = 741  sum = 741
Lookout: convert file name from charset: WINDOWS-1252
checksum = 34584  sum = 34584
checksum = 19127  sum = 19127
Lookout: convert file name from charset: WINDOWS-1252
TNEF: Entering tnef_file_notify()
TNEF: Notifying listener of redäcted.PDF, pos = 0, has data = true, is final = true
LookOut: Entering onTnefStart()
LookOut:    open or save: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2
/var/folders/qx/redacted/T/redäcted.PDF -> [".PDF", (void 0), ".PDF"]
LookOut:    Parent: null
            mMsgUri: imap-message://redacted@imap.redacted/redacted#1234
            requested Part_ID: 2
            Part_ID: 2
            Displayname: redäcted.PDF
            Content-Type: application/binary
            Length: 97475
            URL: file:///var/folders/qx/redacted/T/red%C3%A4cted.PDF
            mimeurl: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2
LookOut: onTnefData position 0  data.len 97475  outstrm [xpconnect wrapped nsIFileOutputStream]
LookOut:    writing 97475bytes to file
LookOut: Entering onTnefEnd()
LookOut:     saving attachment 'undefined'
Lookout: Entering redraw_attachment_view()
LookOut: Entering onStopRequest()
```

note that Umlauts were present in the filename, and have been reasonably preserved in redacted log.

alternative approach:

1. double-click file in Attachment-List, select "Save As ..."
2. select place to save in dialog, click Save
3. don't find file saved.

```
LookOut: Entering save_attachment()
LookOut:    got attachment = ({contentType:"application/binary", name:"red\xE4cted.PDF", uri:"imap-message://redacted@imap.redacted/redacted#1234", isExternalAttachment:true, size:97475, partID:"1.2", url:"imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2", lo_orig_open:(function AttachmentInfo_open()
  {
    if (!this.hasFile)
      return;

    if (this.isEmpty) {
      var prompt = document.getElementById("bundle_messenger")
                           .getString("emptyAttachment");
      msgWindow.promptDialog.alert(null, prompt);
    } else {
      messenger.openAttachment(this.contentType, this.url,
                               encodeURIComponent(this.name),
                               this.uri, this.isExternalAttachment);
    }
  }), open:(function () {

				lookout_lib.open_attachment( this );

			}), lo_orig_save:(function AttachmentInfo_save()
  {
    messenger.saveAttachment(this.contentType, this.url,
                             encodeURIComponent(this.name),
                             this.uri, this.isExternalAttachment);
  }), save:(function (save_dir) {

				lookout_lib.save_attachment( this, save_dir );

			}), parent:{contentType:"application/ms-tnef", name:"winmail.dat", uri:"imap-message://redacted@imap.redacted/redacted#1234", isExternalAttachment:false, size:106536, partID:"1.2", url:"imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat"}, part_id:2})
LookOut:   Parent: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat
         Content-Type: application/binary
         Displayname: redäcted.PDF
         Part_ID: 2
         isExternal: true
         URL: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2
         mMsgUri: imap-message://redacted@imap.redacted/redacted#1234
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
TNEF Key: 0803
checksum = 1  sum = 1
checksum = 232  sum = 232
TNEF: OEM Code Page = 1252
checksum = 2097  sum = 2097
checksum = 43593  sum = 43593
TNEF: Entering tnef_file_notify()
TNEF: Notifying listener of body_part_0.rtf, pos = 0, has data = true, is final = true
LookOut: Entering onTnefStart()
LookOut:    open or save: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.1
LookOut:    Parent: null
            mMsgUri: imap-message://redacted@imap.redacted/redacted#1234
            requested Part_ID: 2
            Part_ID: 1
            Displayname: body_part_0.rtf
            Content-Type: application/rtf
            Length: 14646
            URL: 
            mimeurl: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.1
LookOut: onTnefData position 0  data.len 14646  outstrm null
LookOut: Entering onTnefEnd()
Lookout: Entering redraw_attachment_view()
checksum = 1085  sum = 1085
starting file sub-attachment
checksum = 317  sum = 317
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
LookOut: Entering onDataAvailable()
TNEF: Entering tnef_pack_parse_stream()
checksum = 39746  sum = 39746
checksum = 741  sum = 741
Lookout: convert file name from charset: WINDOWS-1252
checksum = 34584  sum = 34584
checksum = 19127  sum = 19127
Lookout: convert file name from charset: WINDOWS-1252
TNEF: Entering tnef_file_notify()
TNEF: Notifying listener of redäcted.PDF, pos = 0, has data = true, is final = true
LookOut: Entering onTnefStart()
LookOut:    open or save: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2
/var/folders/qx/redacted/T/redäcted.PDF -> [".PDF", (void 0), ".PDF"]
LookOut:    Parent: null
            mMsgUri: imap-message://redacted@imap.redacted/redacted#1234
            requested Part_ID: 2
            Part_ID: 2
            Displayname: redäcted.PDF
            Content-Type: application/binary
            Length: 97475
            URL: file:///var/folders/qx/redacted/T/red%C3%A4cted.PDF
            mimeurl: imap://redacted@imap.redacted:993/fetch%3EUID%3E/redacted%3E1234?part=1.2&filename=winmail.dat.2
LookOut: onTnefData position 0  data.len 97475  outstrm [xpconnect wrapped nsIFileOutputStream]
LookOut:    writing 97475bytes to file
LookOut: Entering onTnefEnd()
LookOut:     saving attachment 'undefined'
Lookout: Entering redraw_attachment_view()
LookOut: Entering onStopRequest()
```

NOTE: upon closer inspection, it seems like the filename might be mangled due to Charset-encoding -conversion? i.e. I found what would be `redäcted.PDF` in one case in the selected folder, in another case in the `/var/folders/qx/redacted/T/` folder .. mostly mangled like `redäcte-14.PDF` or similar.

Disabling Filename character set conversion doesn't make a difference for either case.

**Comment by dugite-code** [comment]
Humm... As far as I understand the special characters in the file name should be converted into characters the filesystem can handle, but then I have only ever used this on windows.

Could you give [Version 1.2.24 Beta 1](https://github.com/TB-throwback/LookOut-fix-version/releases/tag/1.2.24b1) a try. It might solve the `saving attachment 'undefined'` part at least

**Comment by dugite-code** [comment]
Did the beta help at all?
